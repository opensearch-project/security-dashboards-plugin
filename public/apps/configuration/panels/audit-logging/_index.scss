.described-form-group {
  max-width: 1500px;
}

.form-row {
  max-width: 800px;
}
