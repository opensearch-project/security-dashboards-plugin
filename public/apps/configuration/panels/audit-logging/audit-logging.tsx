/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import {
  EuiButton,
  EuiCode,
  EuiDescribedFormGroup,
  EuiFlexGroup,
  EuiFlexItem,
  EuiForm,
  EuiFormRow,
  EuiHorizontalRule,
  EuiLoadingContent,
  EuiPageHeader,
  EuiPanel,
  EuiSpacer,
  EuiSwitch,
  EuiText,
  EuiTitle,
} from '@elastic/eui';
import React, { useContext } from 'react';
import { FormattedMessage } from '@osd/i18n/react';
import { DataSourceOption } from 'src/plugins/data_source_management/public';
import { AppDependencies } from '../../../types';
import { ResourceType } from '../../../../../common';
import { getAuditLogging, updateAuditLogging } from '../../utils/audit-logging-utils';
import { displayBoolean, ExternalLink } from '../../utils/display-utils';
import { buildHashUrl } from '../../utils/url-builder';
import {
  SETTING_GROUPS,
  SUB_URL_FOR_COMPLIANCE_SETTINGS_EDIT,
  SUB_URL_FOR_GENERAL_SETTINGS_EDIT,
} from './constants';
import { AuditLoggingSettings } from './types';
import { ViewSettingGroup } from './view-setting-group';
import { DocLinks } from '../../constants';
import { DataSourceContext } from '../../app-router';
import { SecurityPluginTopNavMenu } from '../../top-nav-menu';
import { AccessErrorComponent } from '../../access-error-component';

interface AuditLoggingProps extends AppDependencies {
  fromType: string;
}

function renderStatusPanel(onSwitchChange: () => void, auditLoggingEnabled: boolean) {
  return (
    <EuiPanel>
      <EuiForm>
        <EuiDescribedFormGroup title={<h3>Storage location</h3>} className="described-form-group">
          <EuiFormRow className="form-row">
            <EuiText color="subdued" grow={false}>
              <FormattedMessage
                id="audit.logs.storageInstruction"
                defaultMessage="Configure the output location and storage types in {opensearchCode}. The default storage location is {internalOpenSearchCode}, which stores the logs in an index on this cluster."
                values={{
                  opensearchCode: <EuiCode>opensearch.yml</EuiCode>,
                  internalOpenSearchCode: <EuiCode>internal_opensearch</EuiCode>,
                }}
              />{' '}
              <ExternalLink href={DocLinks.AuditLogsStorageDoc} />
            </EuiText>
          </EuiFormRow>
        </EuiDescribedFormGroup>

        <EuiDescribedFormGroup
          title={<h3>Enable audit logging</h3>}
          className="described-form-group"
        >
          <EuiFormRow>
            <EuiSwitch
              data-test-subj="audit-logging-enabled-switch"
              name="auditLoggingEnabledSwitch"
              label={displayBoolean(auditLoggingEnabled)}
              checked={auditLoggingEnabled}
              onChange={onSwitchChange}
            />
          </EuiFormRow>
        </EuiDescribedFormGroup>
      </EuiForm>
    </EuiPanel>
  );
}

function renderAccessErrorPanel(loading: boolean, dataSource: DataSourceOption) {
  return (
    <AccessErrorComponent
      loading={loading}
      dataSourceLabel={dataSource && dataSource.label}
      message="You do not have permissions to configure audit logging settings"
    />
  );
}

export function renderGeneralSettings(config: AuditLoggingSettings) {
  return (
    <>
      <ViewSettingGroup config={config} settingGroup={SETTING_GROUPS.LAYER_SETTINGS} />

      <EuiSpacer />

      <ViewSettingGroup config={config} settingGroup={SETTING_GROUPS.ATTRIBUTE_SETTINGS} />

      <EuiSpacer />

      <ViewSettingGroup config={config} settingGroup={SETTING_GROUPS.IGNORE_SETTINGS} />
    </>
  );
}

export function renderComplianceSettings(config: AuditLoggingSettings) {
  return (
    <>
      <ViewSettingGroup
        config={config}
        settingGroup={SETTING_GROUPS.COMPLIANCE_CONFIG_MODE_SETTINGS}
      />

      <EuiSpacer />

      <ViewSettingGroup config={config} settingGroup={SETTING_GROUPS.COMPLIANCE_CONFIG_SETTINGS} />

      <EuiSpacer />

      <ViewSettingGroup config={config} settingGroup={SETTING_GROUPS.COMPLIANCE_SETTINGS_READ} />

      <EuiSpacer />

      <ViewSettingGroup config={config} settingGroup={SETTING_GROUPS.COMPLIANCE_SETTINGS_WRITE} />

      <EuiSpacer />
    </>
  );
}

export function AuditLogging(props: AuditLoggingProps) {
  const [configuration, setConfiguration] = React.useState<AuditLoggingSettings>({});
  const { dataSource, setDataSource } = useContext(DataSourceContext)!;
  const [loading, setLoading] = React.useState(false);
  const [accessErrorFlag, setAccessErrorFlag] = React.useState(false);

  const onSwitchChange = async () => {
    try {
      const updatedConfiguration = { ...configuration };
      updatedConfiguration.enabled = !updatedConfiguration.enabled;

      await updateAuditLogging(props.coreStart.http, updatedConfiguration, dataSource.id);

      setConfiguration(updatedConfiguration);
    } catch (e) {
      console.error(e);
    }
  };

  React.useEffect(() => {
    const fetchData = async () => {
      try {
        setLoading(true);
        const auditLogging = await getAuditLogging(props.coreStart.http, dataSource.id);
        setConfiguration(auditLogging);
        setAccessErrorFlag(false);
      } catch (e) {
        // TODO: switch to better error handling.
        console.log(e);
        // requests with existing credentials but insufficient permissions result in 403, remote data-source requests with non-existing credentials result in 400
        if (e.response && [400, 403].includes(e.response.status)) {
          setAccessErrorFlag(true);
        }
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [props.coreStart.http, props.fromType, dataSource]);

  const statusPanel = renderStatusPanel(onSwitchChange, configuration.enabled || false);

  let content;

  if (accessErrorFlag) {
    content = renderAccessErrorPanel(loading, dataSource);
  } else if (!configuration.enabled) {
    content = statusPanel;
  } else {
    content = (
      <>
        {statusPanel}
        <EuiSpacer />
        <EuiPanel data-test-subj="general-settings">
          <EuiFlexGroup>
            <EuiFlexItem>
              <EuiTitle>
                <h3>General settings</h3>
              </EuiTitle>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiButton
                data-test-subj="general-settings-configure"
                onClick={() => {
                  window.location.href =
                    buildHashUrl(ResourceType.auditLogging) + SUB_URL_FOR_GENERAL_SETTINGS_EDIT;
                }}
              >
                Configure
              </EuiButton>
            </EuiFlexItem>
          </EuiFlexGroup>
          <EuiHorizontalRule margin="m" />
          {renderGeneralSettings(configuration)}
        </EuiPanel>

        <EuiSpacer />

        <EuiPanel>
          <EuiFlexGroup>
            <EuiFlexItem>
              <EuiTitle>
                <h3>Compliance settings</h3>
              </EuiTitle>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiButton
                data-test-subj="compliance-settings-configure"
                onClick={() => {
                  window.location.href =
                    buildHashUrl(ResourceType.auditLogging) + SUB_URL_FOR_COMPLIANCE_SETTINGS_EDIT;
                }}
              >
                Configure
              </EuiButton>
            </EuiFlexItem>
          </EuiFlexGroup>

          <EuiHorizontalRule margin="m" />
          {renderComplianceSettings(configuration)}
        </EuiPanel>
      </>
    );
  }

  return (
    <div className="panel-restrict-width">
      <SecurityPluginTopNavMenu
        {...props}
        dataSourcePickerReadOnly={false}
        setDataSource={setDataSource}
        selectedDataSource={dataSource}
      />
      <EuiPageHeader>
        <EuiTitle size="l">
          <h1>Audit Logging</h1>
        </EuiTitle>
      </EuiPageHeader>
      <EuiSpacer />
      {loading ? <EuiLoadingContent /> : content}
    </div>
  );
}
