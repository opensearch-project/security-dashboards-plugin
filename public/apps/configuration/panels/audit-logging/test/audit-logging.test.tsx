/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { shallow } from 'enzyme';
import { AuditLogging, renderComplianceSettings, renderGeneralSettings } from '../audit-logging';
import React from 'react';
import { EuiSwitch } from '@elastic/eui';
import { buildHashUrl } from '../../../utils/url-builder';
import { ResourceType } from '../../../../../../common';
import {
  SUB_URL_FOR_COMPLIANCE_SETTINGS_EDIT,
  SUB_URL_FOR_GENERAL_SETTINGS_EDIT,
} from '../constants';

jest.mock('../../../utils/audit-logging-utils');
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: jest.fn().mockReturnValue({ dataSource: { id: 'test' }, setDataSource: jest.fn() }), // Mock the useContext hook to return dummy datasource and setdatasource function
}));

// eslint-disable-next-line
const mockAuditLoggingUtils = require('../../../utils/audit-logging-utils');

describe('Audit logs', () => {
  const setState = jest.fn();
  const mockCoreStart = {
    http: 1,
  };

  beforeEach(() => {
    jest.spyOn(React, 'useState').mockRestore();
    jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => [[], setState])
      .mockImplementationOnce(() => [false, jest.fn()]);
    jest.spyOn(React, 'useEffect').mockImplementationOnce((f) => f());
  });

  it('Render disabled', () => {
    const mockAuditLoggingData = {
      enabled: false,
      audit: '',
      compliance: '',
    };

    mockAuditLoggingUtils.getAuditLogging = jest.fn().mockReturnValue(mockAuditLoggingData);

    const component = shallow(
      <AuditLogging coreStart={mockCoreStart as any} navigation={{} as any} />
    );

    const switchFound = component.find(EuiSwitch);

    expect(switchFound.prop('checked')).toBeFalsy();
    expect(switchFound.prop('label')).toBe('Disabled');
  });

  it('Render enabled', (done) => {
    const mockAuditLoggingData = {
      enabled: true,
      audit: '',
      compliance: '',
    };

    mockAuditLoggingUtils.getAuditLogging = jest.fn().mockReturnValue(mockAuditLoggingData);

    shallow(<AuditLogging coreStart={mockCoreStart as any} navigation={{} as any} />);

    process.nextTick(() => {
      expect(mockAuditLoggingUtils.getAuditLogging).toHaveBeenCalledTimes(1);
      expect(setState).toHaveBeenCalledWith(mockAuditLoggingData);

      done();
    });
  });

  it('Render error', (done) => {
    // Hide the error message
    jest.spyOn(console, 'log').mockImplementationOnce(() => {});

    mockAuditLoggingUtils.getAuditLogging = jest.fn().mockImplementationOnce(() => {
      throw Error();
    });

    shallow(<AuditLogging coreStart={mockCoreStart as any} navigation={{} as any} />);

    process.nextTick(() => {
      expect(mockAuditLoggingUtils.getAuditLogging).toHaveBeenCalledTimes(1);
      expect(setState).toBeCalledTimes(0);

      done();
    });
  });

  it('render general settings', () => {
    const generalSettings = renderGeneralSettings({});
    const Component = () => <>{generalSettings}</>;
    const wrapper = shallow(<Component />);
    expect(wrapper).toMatchSnapshot();
  });

  it('render compliance settings', () => {
    const complianceSettings = renderComplianceSettings({});
    const Component = () => <>{complianceSettings}</>;
    const wrapper = shallow(<Component />);
    expect(wrapper).toMatchSnapshot();
  });

  it('audit logging switch change', () => {
    const component = shallow(
      <AuditLogging coreStart={mockCoreStart as any} navigation={{} as any} />
    );
    component.find('[data-test-subj="audit-logging-enabled-switch"]').simulate('change');
    expect(mockAuditLoggingUtils.updateAuditLogging).toHaveBeenCalledTimes(1);
  });

  it('should log error if error occurred while switch change', () => {
    // Hide the error message
    const spy = jest.spyOn(console, 'error').mockImplementationOnce(() => {});

    mockAuditLoggingUtils.updateAuditLogging = jest.fn().mockImplementationOnce(() => {
      throw Error();
    });
    const component = shallow(
      <AuditLogging coreStart={mockCoreStart as any} navigation={{} as any} />
    );
    component.find('[data-test-subj="audit-logging-enabled-switch"]').simulate('change');

    expect(spy).toBeCalled();
    expect(setState).toBeCalledTimes(0);
  });

  it('render when AuditLoggingSettings.enabled is true', () => {
    const auditLoggingSettings = { enabled: true };
    jest.spyOn(React, 'useState').mockRestore();
    jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => [auditLoggingSettings, setState])
      .mockImplementationOnce(() => [false, jest.fn()]);
    const component = shallow(
      <AuditLogging coreStart={mockCoreStart as any} navigation={{} as any} />
    );
    expect(component).toMatchSnapshot();
  });

  it('Click Configure button of general setting section', () => {
    const auditLoggingSettings = { enabled: true };
    jest.spyOn(React, 'useState').mockRestore();
    jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => [auditLoggingSettings, setState])
      .mockImplementationOnce(() => [false, jest.fn()]);
    const component = shallow(
      <AuditLogging coreStart={mockCoreStart as any} navigation={{} as any} />
    );
    component.find('[data-test-subj="general-settings-configure"]').simulate('click');
    expect(window.location.hash).toBe(
      buildHashUrl(ResourceType.auditLogging) + SUB_URL_FOR_GENERAL_SETTINGS_EDIT
    );
  });

  it('Click Configure button of Compliance settings section', () => {
    const auditLoggingSettings = { enabled: true };
    jest.spyOn(React, 'useState').mockRestore();
    jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => [auditLoggingSettings, setState])
      .mockImplementationOnce(() => [false, jest.fn()]);
    const component = shallow(
      <AuditLogging coreStart={mockCoreStart as any} navigation={{} as any} />
    );
    component.find('[data-test-subj="compliance-settings-configure"]').simulate('click');
    expect(window.location.hash).toBe(
      buildHashUrl(ResourceType.auditLogging) + SUB_URL_FOR_COMPLIANCE_SETTINGS_EDIT
    );
  });

  it('should load access error component', () => {
    const auditLoggingSettings = { enabled: true };
    mockAuditLoggingUtils.getAuditLogging = jest
      .fn()
      .mockRejectedValue({ response: { status: 403 } });
    jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => [auditLoggingSettings, setState])
      .mockImplementationOnce(() => [false, jest.fn()])
      .mockImplementationOnce(() => [true, jest.fn()]);
    const component = shallow(
      <AuditLogging coreStart={mockCoreStart as any} navigation={{} as any} />
    );
    expect(component).toMatchSnapshot();
  });
});
