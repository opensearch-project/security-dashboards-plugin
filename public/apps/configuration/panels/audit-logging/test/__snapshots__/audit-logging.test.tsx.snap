// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Audit logs render compliance settings 1`] = `
<Fragment>
  <ViewSettingGroup
    config={Object {}}
    settingGroup={
      Object {
        "settings": Array [
          Object {
            "description": "Enable or disable compliance logging.",
            "path": "compliance.enabled",
            "title": "Compliance logging",
            "type": "bool",
          },
        ],
        "title": "Compliance mode",
      }
    }
  />
  <EuiSpacer />
  <ViewSettingGroup
    config={Object {}}
    settingGroup={
      Object {
        "settings": Array [
          Object {
            "description": "Enable or disable logging of events on internal security index.",
            "path": "compliance.internal_config",
            "title": "Internal config logging",
            "type": "bool",
          },
          Object {
            "description": "Enable or disable logging of external configuration.",
            "path": "compliance.external_config",
            "title": "External config logging",
            "type": "bool",
          },
        ],
        "title": "Config",
      }
    }
  />
  <EuiSpacer />
  <ViewSettingGroup
    config={Object {}}
    settingGroup={
      Object {
        "settings": Array [
          Object {
            "description": "Do not log any document fields. Log only metadata of the document.",
            "path": "compliance.read_metadata_only",
            "title": "Read metadata",
            "type": "bool",
          },
          Object {
            "description": "Users to ignore during auditing.",
            "path": "compliance.read_ignore_users",
            "placeHolder": "Add users or user patterns",
            "title": "Ignored users",
            "type": "array",
          },
          Object {
            "code": "{
  \\"index-name-pattern\\": [\\"field-name-pattern\\"],
  \\"logs*\\": [\\"message\\"],
  \\"twitter\\": [\\"id\\", \\"user*\\"]
}",
            "description": "List the indices and fields to watch during read events. Adding watched fields will generate one log per document access and could result in significant overhead. Sample data content:",
            "error": "Invalid content. Please check sample data content.",
            "path": "compliance.read_watched_fields",
            "title": "Watched fields",
            "type": "map",
          },
        ],
        "title": "Read",
      }
    }
  />
  <EuiSpacer />
  <ViewSettingGroup
    config={Object {}}
    settingGroup={
      Object {
        "settings": Array [
          Object {
            "description": "Do not log any document content. Log only metadata of the document.",
            "path": "compliance.write_metadata_only",
            "title": "Write metadata",
            "type": "bool",
          },
          Object {
            "description": "Log diffs for document updates.",
            "path": "compliance.write_log_diffs",
            "title": "Log diffs",
            "type": "bool",
          },
          Object {
            "description": "Users to ignore during auditing.",
            "path": "compliance.write_ignore_users",
            "placeHolder": "Add users or user patterns",
            "title": "Ignored users",
            "type": "array",
          },
          Object {
            "description": "List the indices to watch during write events. Adding watched indices will generate one log per document access and could result in significant overhead.",
            "path": "compliance.write_watched_indices",
            "placeHolder": "Add indices",
            "title": "Watch indices",
            "type": "array",
          },
        ],
        "title": "Write",
      }
    }
  />
  <EuiSpacer />
</Fragment>
`;

exports[`Audit logs render general settings 1`] = `
<Fragment>
  <ViewSettingGroup
    config={Object {}}
    settingGroup={
      Object {
        "settings": Array [
          Object {
            "description": "Enable or disable auditing events that happen on the REST layer.",
            "path": "audit.enable_rest",
            "title": "REST layer",
            "type": "bool",
          },
          Object {
            "description": "Specify audit categories which must be ignored on the REST layer. Modifying these could result in significant overhead.",
            "options": Array [
              "BAD_HEADERS",
              "FAILED_LOGIN",
              "MISSING_PRIVILEGES",
              "GRANTED_PRIVILEGES",
              "SSL_EXCEPTION",
              "AUTHENTICATED",
            ],
            "path": "audit.disabled_rest_categories",
            "placeHolder": "Select categories",
            "title": "REST disabled categories",
            "type": "array",
          },
          Object {
            "description": "Enable or disable auditing events that happen on the transport layer.",
            "path": "audit.enable_transport",
            "title": "Transport layer",
            "type": "bool",
          },
          Object {
            "description": "Specify audit categories which must be ignored on the transport layer. Modifying these could result in significant overhead.",
            "options": Array [
              "BAD_HEADERS",
              "FAILED_LOGIN",
              "GRANTED_PRIVILEGES",
              "INDEX_EVENT",
              "MISSING_PRIVILEGES",
              "SSL_EXCEPTION",
              "OPENDISTRO_SECURITY_INDEX_ATTEMPT",
              "AUTHENTICATED",
            ],
            "path": "audit.disabled_transport_categories",
            "placeHolder": "Select categories",
            "title": "Transport disabled categories",
            "type": "array",
          },
        ],
        "title": "Layer settings",
      }
    }
  />
  <EuiSpacer />
  <ViewSettingGroup
    config={Object {}}
    settingGroup={
      Object {
        "settings": Array [
          Object {
            "description": "Resolve bulk requests during auditing of requests. Enabling this will generate a log for each document request which could result in significant overhead.",
            "path": "audit.resolve_bulk_requests",
            "title": "Bulk requests",
            "type": "bool",
          },
          Object {
            "description": "Include request body during auditing of requests.",
            "path": "audit.log_request_body",
            "title": "Request body",
            "type": "bool",
          },
          Object {
            "description": "Resolve indices during auditing of requests.",
            "path": "audit.resolve_indices",
            "title": "Resolve indices",
            "type": "bool",
          },
          Object {
            "description": "Exclude sensitive headers during auditing. (e.g. authorization header)",
            "path": "audit.exclude_sensitive_headers",
            "title": "Sensitive headers",
            "type": "bool",
          },
        ],
        "title": "Attribute settings",
      }
    }
  />
  <EuiSpacer />
  <ViewSettingGroup
    config={Object {}}
    settingGroup={
      Object {
        "settings": Array [
          Object {
            "description": "Users to ignore during auditing. Changing the defaults could result in significant overhead.",
            "path": "audit.ignore_users",
            "placeHolder": "Add users or user patterns",
            "title": "Ignored users",
            "type": "array",
          },
          Object {
            "description": "Request patterns to ignore during auditing.",
            "path": "audit.ignore_requests",
            "placeHolder": "Add request patterns",
            "title": "Ignored requests",
            "type": "array",
          },
        ],
        "title": "Ignore settings",
      }
    }
  />
</Fragment>
`;

exports[`Audit logs render when AuditLoggingSettings.enabled is true 1`] = `
<div
  className="panel-restrict-width"
>
  <Memo()
    coreStart={
      Object {
        "http": 1,
      }
    }
    dataSourcePickerReadOnly={false}
    navigation={Object {}}
    selectedDataSource={
      Object {
        "id": "test",
      }
    }
    setDataSource={[MockFunction]}
  />
  <EuiPageHeader>
    <EuiTitle
      size="l"
    >
      <h1>
        Audit Logging
      </h1>
    </EuiTitle>
  </EuiPageHeader>
  <EuiSpacer />
  <EuiPanel>
    <EuiForm>
      <EuiDescribedFormGroup
        className="described-form-group"
        title={
          <h3>
            Storage location
          </h3>
        }
      >
        <EuiFormRow
          className="form-row"
          describedByIds={Array []}
          display="row"
          fullWidth={false}
          hasChildLabel={true}
          hasEmptyLabelSpace={false}
          labelType="label"
        >
          <EuiText
            color="subdued"
            grow={false}
          >
            <FormattedMessage
              defaultMessage="Configure the output location and storage types in {opensearchCode}. The default storage location is {internalOpenSearchCode}, which stores the logs in an index on this cluster."
              id="audit.logs.storageInstruction"
              values={
                Object {
                  "internalOpenSearchCode": <EuiCode>
                    internal_opensearch
                  </EuiCode>,
                  "opensearchCode": <EuiCode>
                    opensearch.yml
                  </EuiCode>,
                }
              }
            />
             
            <ExternalLink
              href="https://opensearch.org/docs/latest/security-plugin/audit-logs/storage-types/"
            />
          </EuiText>
        </EuiFormRow>
      </EuiDescribedFormGroup>
      <EuiDescribedFormGroup
        className="described-form-group"
        title={
          <h3>
            Enable audit logging
          </h3>
        }
      >
        <EuiFormRow
          describedByIds={Array []}
          display="row"
          fullWidth={false}
          hasChildLabel={true}
          hasEmptyLabelSpace={false}
          labelType="label"
        >
          <EuiSwitch
            checked={true}
            data-test-subj="audit-logging-enabled-switch"
            label="Enabled"
            name="auditLoggingEnabledSwitch"
            onChange={[Function]}
          />
        </EuiFormRow>
      </EuiDescribedFormGroup>
    </EuiForm>
  </EuiPanel>
  <EuiSpacer />
  <EuiPanel
    data-test-subj="general-settings"
  >
    <EuiFlexGroup>
      <EuiFlexItem>
        <EuiTitle>
          <h3>
            General settings
          </h3>
        </EuiTitle>
      </EuiFlexItem>
      <EuiFlexItem
        grow={false}
      >
        <EuiButton
          data-test-subj="general-settings-configure"
          onClick={[Function]}
        >
          Configure
        </EuiButton>
      </EuiFlexItem>
    </EuiFlexGroup>
    <EuiHorizontalRule
      margin="m"
    />
    <ViewSettingGroup
      config={
        Object {
          "enabled": true,
        }
      }
      settingGroup={
        Object {
          "settings": Array [
            Object {
              "description": "Enable or disable auditing events that happen on the REST layer.",
              "path": "audit.enable_rest",
              "title": "REST layer",
              "type": "bool",
            },
            Object {
              "description": "Specify audit categories which must be ignored on the REST layer. Modifying these could result in significant overhead.",
              "options": Array [
                "BAD_HEADERS",
                "FAILED_LOGIN",
                "MISSING_PRIVILEGES",
                "GRANTED_PRIVILEGES",
                "SSL_EXCEPTION",
                "AUTHENTICATED",
              ],
              "path": "audit.disabled_rest_categories",
              "placeHolder": "Select categories",
              "title": "REST disabled categories",
              "type": "array",
            },
            Object {
              "description": "Enable or disable auditing events that happen on the transport layer.",
              "path": "audit.enable_transport",
              "title": "Transport layer",
              "type": "bool",
            },
            Object {
              "description": "Specify audit categories which must be ignored on the transport layer. Modifying these could result in significant overhead.",
              "options": Array [
                "BAD_HEADERS",
                "FAILED_LOGIN",
                "GRANTED_PRIVILEGES",
                "INDEX_EVENT",
                "MISSING_PRIVILEGES",
                "SSL_EXCEPTION",
                "OPENDISTRO_SECURITY_INDEX_ATTEMPT",
                "AUTHENTICATED",
              ],
              "path": "audit.disabled_transport_categories",
              "placeHolder": "Select categories",
              "title": "Transport disabled categories",
              "type": "array",
            },
          ],
          "title": "Layer settings",
        }
      }
    />
    <EuiSpacer />
    <ViewSettingGroup
      config={
        Object {
          "enabled": true,
        }
      }
      settingGroup={
        Object {
          "settings": Array [
            Object {
              "description": "Resolve bulk requests during auditing of requests. Enabling this will generate a log for each document request which could result in significant overhead.",
              "path": "audit.resolve_bulk_requests",
              "title": "Bulk requests",
              "type": "bool",
            },
            Object {
              "description": "Include request body during auditing of requests.",
              "path": "audit.log_request_body",
              "title": "Request body",
              "type": "bool",
            },
            Object {
              "description": "Resolve indices during auditing of requests.",
              "path": "audit.resolve_indices",
              "title": "Resolve indices",
              "type": "bool",
            },
            Object {
              "description": "Exclude sensitive headers during auditing. (e.g. authorization header)",
              "path": "audit.exclude_sensitive_headers",
              "title": "Sensitive headers",
              "type": "bool",
            },
          ],
          "title": "Attribute settings",
        }
      }
    />
    <EuiSpacer />
    <ViewSettingGroup
      config={
        Object {
          "enabled": true,
        }
      }
      settingGroup={
        Object {
          "settings": Array [
            Object {
              "description": "Users to ignore during auditing. Changing the defaults could result in significant overhead.",
              "path": "audit.ignore_users",
              "placeHolder": "Add users or user patterns",
              "title": "Ignored users",
              "type": "array",
            },
            Object {
              "description": "Request patterns to ignore during auditing.",
              "path": "audit.ignore_requests",
              "placeHolder": "Add request patterns",
              "title": "Ignored requests",
              "type": "array",
            },
          ],
          "title": "Ignore settings",
        }
      }
    />
  </EuiPanel>
  <EuiSpacer />
  <EuiPanel>
    <EuiFlexGroup>
      <EuiFlexItem>
        <EuiTitle>
          <h3>
            Compliance settings
          </h3>
        </EuiTitle>
      </EuiFlexItem>
      <EuiFlexItem
        grow={false}
      >
        <EuiButton
          data-test-subj="compliance-settings-configure"
          onClick={[Function]}
        >
          Configure
        </EuiButton>
      </EuiFlexItem>
    </EuiFlexGroup>
    <EuiHorizontalRule
      margin="m"
    />
    <ViewSettingGroup
      config={
        Object {
          "enabled": true,
        }
      }
      settingGroup={
        Object {
          "settings": Array [
            Object {
              "description": "Enable or disable compliance logging.",
              "path": "compliance.enabled",
              "title": "Compliance logging",
              "type": "bool",
            },
          ],
          "title": "Compliance mode",
        }
      }
    />
    <EuiSpacer />
    <ViewSettingGroup
      config={
        Object {
          "enabled": true,
        }
      }
      settingGroup={
        Object {
          "settings": Array [
            Object {
              "description": "Enable or disable logging of events on internal security index.",
              "path": "compliance.internal_config",
              "title": "Internal config logging",
              "type": "bool",
            },
            Object {
              "description": "Enable or disable logging of external configuration.",
              "path": "compliance.external_config",
              "title": "External config logging",
              "type": "bool",
            },
          ],
          "title": "Config",
        }
      }
    />
    <EuiSpacer />
    <ViewSettingGroup
      config={
        Object {
          "enabled": true,
        }
      }
      settingGroup={
        Object {
          "settings": Array [
            Object {
              "description": "Do not log any document fields. Log only metadata of the document.",
              "path": "compliance.read_metadata_only",
              "title": "Read metadata",
              "type": "bool",
            },
            Object {
              "description": "Users to ignore during auditing.",
              "path": "compliance.read_ignore_users",
              "placeHolder": "Add users or user patterns",
              "title": "Ignored users",
              "type": "array",
            },
            Object {
              "code": "{
  \\"index-name-pattern\\": [\\"field-name-pattern\\"],
  \\"logs*\\": [\\"message\\"],
  \\"twitter\\": [\\"id\\", \\"user*\\"]
}",
              "description": "List the indices and fields to watch during read events. Adding watched fields will generate one log per document access and could result in significant overhead. Sample data content:",
              "error": "Invalid content. Please check sample data content.",
              "path": "compliance.read_watched_fields",
              "title": "Watched fields",
              "type": "map",
            },
          ],
          "title": "Read",
        }
      }
    />
    <EuiSpacer />
    <ViewSettingGroup
      config={
        Object {
          "enabled": true,
        }
      }
      settingGroup={
        Object {
          "settings": Array [
            Object {
              "description": "Do not log any document content. Log only metadata of the document.",
              "path": "compliance.write_metadata_only",
              "title": "Write metadata",
              "type": "bool",
            },
            Object {
              "description": "Log diffs for document updates.",
              "path": "compliance.write_log_diffs",
              "title": "Log diffs",
              "type": "bool",
            },
            Object {
              "description": "Users to ignore during auditing.",
              "path": "compliance.write_ignore_users",
              "placeHolder": "Add users or user patterns",
              "title": "Ignored users",
              "type": "array",
            },
            Object {
              "description": "List the indices to watch during write events. Adding watched indices will generate one log per document access and could result in significant overhead.",
              "path": "compliance.write_watched_indices",
              "placeHolder": "Add indices",
              "title": "Watch indices",
              "type": "array",
            },
          ],
          "title": "Write",
        }
      }
    />
    <EuiSpacer />
  </EuiPanel>
</div>
`;

exports[`Audit logs should load access error component 1`] = `
<div
  className="panel-restrict-width"
>
  <Memo()
    coreStart={
      Object {
        "http": 1,
      }
    }
    dataSourcePickerReadOnly={false}
    navigation={Object {}}
    selectedDataSource={
      Object {
        "id": "test",
      }
    }
    setDataSource={[MockFunction]}
  />
  <EuiPageHeader>
    <EuiTitle
      size="l"
    >
      <h1>
        Audit Logging
      </h1>
    </EuiTitle>
  </EuiPageHeader>
  <EuiSpacer />
  <AccessErrorComponent
    loading={false}
    message="You do not have permissions to configure audit logging settings"
  />
</div>
`;
